examples/mamba/convert_hf_to_nanotron.py

```diff
@@ -0,0 +1,342 @@
+#             param_shard_ref = param_ref
+#             if isinstance(param, NanotronParameter) and param.is_sharded and tp_size > 1:
+#                 dim = next(index for index, (dim1, dim2) in enumerate(zip(param.shape, param_ref.shape)) if dim1 != dim2)
+#                 param_shard_ref = _split_weight(param_ref, dim)
+            
+#             if "in_proj" in name_ref:
+#                 # Don't check this weight as we changed it manually (interleaved)
+#                 excluded += 1
+#                 continue
+            
+#             torch.testing.assert_close(param_shard_ref, param, rtol=1e-10, atol=1e-10)
+#         except AssertionError as e:
+#             log_rank(f"{name_ref} and {name} are not equal. {e}", logger=logger, level=logging.INFO, rank=0)
+#             fail += 1
+    
+#     log_rank(f"{excluded}/{total} parameters were not sanity check (interleaved)", logger=logger, level=logging.INFO, rank=0)
+#     log_rank(f"{fail}/{total} parameters are not equal", logger=logger, level=logging.INFO, rank=0)
+    
+#     if fail > 0:
+#         raise AssertionError("Some parameters are not equal")
+
+def get_weight_from_hf(
+    name: str,
+    ref_module_state_dict: Dict[str, torch.Tensor],
+    ref_module: MambaLMHeadModel,
+    nanotron_to_hf: Dict[str, str],
+    get_grad: bool = False
+) -> torch.Tensor:
+    """From our brrr implementation, we get the equivalent tensor in transformers implementation"""    
+  
+    def _interleave_pattern(N):
+        """
+        interleave_pattern(4) -> [0, 2, 1, 3]
+        interleave_pattern(8) -> [0, 4, 1, 5, 2, 6, 3, 7]
+        """
+        assert N % 2 == 0, "N must be even"
+        pattern = []
+        for i in range(N // 2):
+            pattern.append(i)
+            pattern.append(i + N // 2)
+        return pattern
+  
+    hf_name = nanotron_to_hf[name]
+    
+    if get_grad is False:
+        def _get_tensor(path: str):
+            return ref_module_state_dict[path]
+    else:
+        def _get_tensor(path: str):
+            param = ref_module.get_parameter(path)
+            return param.grad
+    
+    param = _get_tensor(hf_name)
+    
+    if "in_proj" in hf_name:
+        # In Nanotron, we do tensor parallel column so weight need to be split in the column dimension (i.e: xz.view(...))
+        # However, the HF weights was trained such that it expected xz.chunk(...) to split the tensor in the row dimension
+        # Thus, we need to interleaved the HF weights to make it compatible with Nanotron
+        log_rank(f"Interleaving {hf_name} to make it compatible with Nanotron", logger=logger, level=logging.INFO, rank=0)
+        param = param[_interleave_pattern(param.shape[0]), :]
+        
+    return param
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Convert HF weights from states-space repo to brrr weights")
+    parser.add_argument("--model", type=str, default="130M", help="130M | 370M | 790M | 1.4B | 2.8B")
```

Comment:
I think it's better to have the entire model name here and not just 130M
The assert below can check if it's a states space mamba model for example